akka-persistence/src/test/scala/akka/persistence/journal/SteppingInmemJournal.scala

```diff
@@ -0,0 +1,136 @@
+import scala.concurrent.{ Future, Promise }
+import scala.util.Try
+
+object SteppingInmemJournal {
+
+  /** allow the journal to do one operation */
+  case object Token
+  case object TokenConsumed
+
+  def config(instanceId: String): Config =
+    ConfigFactory.parseString(
+      s"""
+        |akka.persistence.journal.stepping-inmem.class=akka.persistence.journal.SteppingInmemJournal
+        |akka.persistence.journal.plugin = "akka.persistence.journal.stepping-inmem"
+        |akka.persistence.journal.stepping-inmem.instance-id = "$instanceId"
+      """.stripMargin)
+
+  // keep it in a thread safe:d global so that tests can get their
+  // hand on the actor ref and send Steps to it
+  private[this] var _current: Map[String, ActorRef] = Map()
+
+  // shhh don't tell anyone I sinn-croniz-ed
+  /** get the actor ref to the journal for a given instance id, throws exception if not found */
+  def getRef(instanceId: String): ActorRef = synchronized(_current(instanceId))
+
+  private def putRef(instanceId: String, instance: ActorRef): Unit = synchronized {
+    _current = _current + (instanceId -> instance)
+  }
+  private def remove(instanceId: String): Unit = synchronized(
+    _current -= instanceId)
+}
+
+/**
+ * An in memory journal that will not complete any persists or persistAsyncs until it gets tokens
+ * to trigger those steps. Allows for tests that need to deterministically trigger the callbacks
+ * intermixed with receiving messages.
+ *
+ * Configure your actor system using {{{SteppingInMemJournal.config}}} and then access
+ * it using {{{SteppingInmemJournal.getRef(String)}}}, send it {{{SteppingInmemJournal.Token}}}s to
+ * allow one journal operation to complete.
+ */
+final class SteppingInmemJournal extends InmemJournal {
+
+  import SteppingInmemJournal._
+  import context.dispatcher
+
+  val instanceId = context.system.settings.config.getString("akka.persistence.journal.stepping-inmem.instance-id")
+
+  var queuedOps: Seq[() ⇒ Future[Unit]] = Seq.empty
+  var queuedTokenRecipients = List.empty[ActorRef]
+
+  override def receivePluginInternal = super.receivePluginInternal orElse {
+    case Token if queuedOps.isEmpty ⇒ queuedTokenRecipients = queuedTokenRecipients :+ sender()
+    case Token ⇒
+      val op +: rest = queuedOps
+      queuedOps = rest
+      val tokenConsumer = sender()
+      op().onComplete(_ ⇒
+        tokenConsumer ! TokenConsumed)
+  }
+
+  override def preStart(): Unit = {
+    SteppingInmemJournal.putRef(instanceId, self)
+    super.preStart()
+  }
+
+  override def postStop(): Unit = {
+    super.postStop()
+    SteppingInmemJournal.remove(instanceId)
+  }
+
+  override def asyncWriteMessages(messages: Seq[AtomicWrite]): Future[Seq[Try[Unit]]] = {
+    val futures = messages.map { message ⇒
+      val promise = Promise[Try[Unit]]()
+      val future = promise.future
+      doOrEnqueue { () ⇒
+        promise.completeWith(super.asyncWriteMessages(Seq(message)).map(_.head))
+        future.map(_ ⇒ ())
```

Comment:
why is this needed? It's not returned AFAICS; doOrEnqueue can be `... => Future[T]` then it'll compile